File: bindings/julia/src/Dns.jl
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# Julia bindings for the proven-dns protocol (DNS server).
#
# Wraps the C-ABI functions from protocols/proven-dns/ffi/zig/src/dns.zig
# via ccall into libproven_dns.so.

module Dns

using ..ProvenServers: check_status, check_slot, SlotId

export DnsState, DnssecState, Rcode, RecordType, RecordClass, DnssecAlgorithm,
       abi_version, create_context, destroy_context, get_state,
       parse_query, begin_lookup, begin_response, set_rcode,
       enable_dnssec, sign_response, validate_dnssec,
       can_transition, can_dnssec_transition

const LIB = "libproven_dns"

# --------------------------------------------------------------------------
# Enumeration types matching Idris2 ABI
# --------------------------------------------------------------------------

"""DNS query lifecycle states matching DnsState in dns.zig."""
@enum DnsState::UInt8 begin
    STATE_IDLE              = 0
    STATE_QUERY_RECEIVED    = 1
    STATE_LOOKUP            = 2
    STATE_RESPONSE_BUILDING = 3
    STATE_SENT              = 4
end

"""DNSSEC processing states."""
@enum DnssecState::UInt8 begin
    DNSSEC_DISABLED   = 0
    DNSSEC_ENABLED    = 1
    DNSSEC_KEY_LOADED = 2
    DNSSEC_SIGNED     = 3
    DNSSEC_VALIDATED  = 4
end

"""DNS response codes."""
@enum Rcode::UInt8 begin
    RCODE_NOERROR  = 0
    RCODE_FORMERR  = 1
    RCODE_SERVFAIL = 2
    RCODE_NXDOMAIN = 3
    RCODE_NOTIMP   = 4
    RCODE_REFUSED  = 5
end

"""DNS record types (subset)."""
@enum RecordType::UInt8 begin
    RTYPE_A     = 0
    RTYPE_AAAA  = 1
    RTYPE_CNAME = 2
    RTYPE_MX    = 3
    RTYPE_NS    = 4
    RTYPE_PTR   = 5
    RTYPE_SOA   = 6
    RTYPE_SRV   = 7
    RTYPE_TXT   = 8
end

"""DNS record class."""
@enum RecordClass::UInt8 begin
    CLASS_IN  = 0
    CLASS_CH  = 1
    CLASS_HS  = 2
    CLASS_ANY = 3
end

"""DNSSEC algorithm identifiers."""
@enum DnssecAlgorithm::UInt8 begin
    ALGO_RSASHA256  = 0
    ALGO_RSASHA512  = 1
    ALGO_ECDSAP256  = 2
    ALGO_ECDSAP384  = 3
    ALGO_ED25519    = 4
end

# --------------------------------------------------------------------------
# ccall declarations
# --------------------------------------------------------------------------

"""Return the ABI version of the linked libproven_dns."""
function abi_version()::UInt32
    ccall((:dns_abi_version, LIB), UInt32, ())
end

"""
    create_context() -> SlotId

Create a new DNS context. Throws on pool exhaustion.
"""
function create_context()::SlotId
    check_slot(ccall((:dns_create_context, LIB), Cint, ()))
end

"""
    destroy_context(slot::SlotId)

Release the given DNS context slot.
"""
function destroy_context(slot::SlotId)::Nothing
    ccall((:dns_destroy_context, LIB), Cvoid, (Cint,), slot)
    nothing
end

"""
    get_state(slot::SlotId) -> DnsState

Get the current DNS lifecycle state.
"""
function get_state(slot::SlotId)::DnsState
    DnsState(ccall((:dns_state, LIB), UInt8, (Cint,), slot))
end

"""
    parse_query(slot::SlotId, data::Vector{UInt8})

Parse a DNS query from raw bytes. Throws on invalid state.
"""
function parse_query(slot::SlotId, data::Vector{UInt8})::Nothing
    raw = ccall((:dns_parse_query, LIB), UInt8,
                (Cint, Ptr{UInt8}, UInt16),
                slot, data, UInt16(length(data)))
    check_status(raw)
end

"""
    begin_lookup(slot::SlotId)

Transition to lookup phase. Throws on invalid state.
"""
function begin_lookup(slot::SlotId)::Nothing
    check_status(ccall((:dns_begin_lookup, LIB), UInt8, (Cint,), slot))
end

"""
    begin_response(slot::SlotId)

Transition to response building phase. Throws on invalid state.
"""
function begin_response(slot::SlotId)::Nothing
    check_status(ccall((:dns_begin_response, LIB), UInt8, (Cint,), slot))
end

"""
    set_rcode(slot::SlotId, code::Rcode)

Set the DNS response code. Throws on invalid state.
"""
function set_rcode(slot::SlotId, code::Rcode)::Nothing
    check_status(ccall((:dns_set_rcode, LIB), UInt8,
                       (Cint, UInt8), slot, UInt8(code)))
end

"""
    enable_dnssec(slot::SlotId)

Enable DNSSEC for the context. Throws on invalid state.
"""
function enable_dnssec(slot::SlotId)::Nothing
    check_status(ccall((:dns_enable_dnssec, LIB), UInt8, (Cint,), slot))
end

"""
    sign_response(slot::SlotId)

Sign the DNS response with DNSSEC. Throws on invalid state.
"""
function sign_response(slot::SlotId)::Nothing
    check_status(ccall((:dns_sign_response, LIB), UInt8, (Cint,), slot))
end

"""
    validate_dnssec(slot::SlotId)

Validate DNSSEC signatures. Throws on invalid state.
"""
function validate_dnssec(slot::SlotId)::Nothing
    check_status(ccall((:dns_validate_dnssec, LIB), UInt8, (Cint,), slot))
end

"""
    can_transition(from::DnsState, to::DnsState) -> Bool

Check whether a DNS lifecycle transition is valid.
"""
function can_transition(from::DnsState, to::DnsState)::Bool
    ccall((:dns_can_transition, LIB), UInt8,
          (UInt8, UInt8), UInt8(from), UInt8(to)) == 0x01
end

"""
    can_dnssec_transition(from::DnssecState, to::DnssecState) -> Bool

Check whether a DNSSEC state transition is valid.
"""
function can_dnssec_transition(from::DnssecState, to::DnssecState)::Bool
    ccall((:dns_can_dnssec_transition, LIB), UInt8,
          (UInt8, UInt8), UInt8(from), UInt8(to)) == 0x01
end

end # module Dns

File: bindings/javascript/src/dns.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// JavaScript bindings for the proven-dns Zig FFI.

import { checkSlot, checkStatus } from "./error.js";
import { loadLibrary } from "./ffi.js";

/** @readonly @enum {number} DNS query lifecycle states. */
export const DnsState = Object.freeze({
    IDLE: 0, QUERY_RECEIVED: 1, LOOKUP: 2, RESPONSE_BUILDING: 3, SENT: 4,
});

/** @readonly @enum {number} DNSSEC states. */
export const DnssecState = Object.freeze({
    DISABLED: 0, ENABLED: 1, KEY_LOADED: 2, VALIDATED: 3,
});

/** @readonly @enum {number} DNSSEC signing algorithms. */
export const DnssecAlgorithm = Object.freeze({
    RSA_SHA256: 0, RSA_SHA512: 1, ECDSA_P256_SHA256: 2,
    ECDSA_P384_SHA384: 3, ED25519: 4,
});

/** @type {object|null} */
let _lib = null;

/** @param {object} [options] */
export async function init(options) {
    if (!_lib) _lib = await loadLibrary("dns", options);
}

function lib() {
    if (!_lib) throw new Error("dns: call init() before using the module");
    return _lib;
}

/**
 * DNS query/response context wrapping a Zig FFI slot.
 */
export class DnsContext {
    /** @param {number} slot */
    constructor(slot) {
        this._slot = slot;
        this._destroyed = false;
    }

    static create() {
        return new DnsContext(checkSlot(lib().dns_create_context()));
    }

    destroy() {
        if (!this._destroyed) {
            lib().dns_destroy_context(this._slot);
            this._destroyed = true;
        }
    }

    /** @returns {number|null} DnsState tag. */
    state() {
        const tag = lib().dns_state(this._slot);
        return tag <= 4 ? tag : null;
    }

    /** @returns {number|null} DnssecState tag. */
    dnssecState() {
        const tag = lib().dns_dnssec_state(this._slot);
        return tag <= 3 ? tag : null;
    }

    /** @returns {number} Response code tag. */
    rcode() { return lib().dns_rcode(this._slot); }

    /** @returns {number} */ answerCount() { return lib().dns_answer_count(this._slot); }
    /** @returns {number} */ authorityCount() { return lib().dns_authority_count(this._slot); }
    /** @returns {number} */ additionalCount() { return lib().dns_additional_count(this._slot); }
    /** @returns {number} */ queryRtype() { return lib().dns_query_rtype(this._slot); }
    /** @returns {number} */ queryClass() { return lib().dns_query_class(this._slot); }

    /** @param {Uint8Array} data */
    parseQuery(data) {
        checkStatus(lib().dns_parse_query(this._slot, data, data.length));
    }

    beginLookup() { checkStatus(lib().dns_begin_lookup(this._slot)); }
    beginResponse() { checkStatus(lib().dns_begin_response(this._slot)); }

    /**
     * @param {number} rtype
     * @param {number} rclass
     * @param {number} ttl
     * @param {Uint8Array} rdata
     */
    addAnswer(rtype, rclass, ttl, rdata) {
        checkStatus(lib().dns_add_answer(this._slot, rtype, rclass, ttl, rdata, rdata.length));
    }

    /** @param {number} rtype @param {number} rclass @param {number} ttl @param {Uint8Array} rdata */
    addAuthority(rtype, rclass, ttl, rdata) {
        checkStatus(lib().dns_add_authority(this._slot, rtype, rclass, ttl, rdata, rdata.length));
    }

    /** @param {number} rtype @param {number} rclass @param {number} ttl @param {Uint8Array} rdata */
    addAdditional(rtype, rclass, ttl, rdata) {
        checkStatus(lib().dns_add_additional(this._slot, rtype, rclass, ttl, rdata, rdata.length));
    }

    /** @param {number} rcodeTag */
    setRcode(rcodeTag) { checkStatus(lib().dns_set_rcode(this._slot, rcodeTag)); }

    /**
     * Build the DNS response. Transitions ResponseBuilding -> Sent.
     * @param {number} [maxLen=512]
     * @returns {Uint8Array} Serialized DNS response.
     */
    buildResponse(maxLen = 512) {
        const buf = new Uint8Array(maxLen);
        const outLen = new Uint16Array(1);
        checkStatus(lib().dns_build_response(this._slot, buf, outLen));
        return buf.subarray(0, outLen[0]);
    }

    enableDnssec() { checkStatus(lib().dns_enable_dnssec(this._slot)); }

    /** @param {number} algo - DnssecAlgorithm tag. */
    loadDnssecKey(algo) { checkStatus(lib().dns_load_dnssec_key(this._slot, algo)); }

    signResponse() { checkStatus(lib().dns_sign_response(this._slot)); }

    /** @returns {boolean} */
    validateDnssec() { return lib().dns_validate_dnssec(this._slot) === 0; }
}

/** @returns {number} */
export function abiVersion() { return lib().dns_abi_version(); }

/** @param {number} from @param {number} to @returns {boolean} */
export function canTransition(from, to) {
    return lib().dns_can_transition(from, to) === 1;
}

/** @param {number} from @param {number} to @returns {boolean} */
export function canDnssecTransition(from, to) {
    return lib().dns_can_dnssec_transition(from, to) === 1;
}
